### PR TITLE
Added config agent driver for ACI with ASR

### PR DESCRIPTION
**Pull request by tbachman**
This adds the configuration agent driver for an ASR
when used with an ACI-enabled fabric.

Signed-off-by: Thomas Bachman tbachman@yahoo.com

### CODE REVIEW

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            try:
+                ip = port['hosting_info']['gateway_ip']
+                return ip
+            except KeyError as e:
+                params = {'key': e}
+                raise cfg_exc.DriverExpectedKeyNotSetException(**params)
+
+    def _get_interface_next_hop_from_hosting_port(self, port):
+        """
+        Extract the underlying subinterface IP and mask for a port
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py]: the comment is wrong there.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        try:
+            ip = port['hosting_info']['next_hop']
+            return ip
+        except KeyError as e:
+            params = {'key': e}
+            raise cfg_exc.DriverExpectedKeyNotSetException(**params)
+
+    def _get_interface_subnet_from_hosting_port(self, port, is_external=False):
+        """
+        Extract the underlying subinterface IP and mask for a port
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py]: the comment is wrong there.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+            out_itfc = self._get_interface_name_from_hosting_port(port)
+            ip = netaddr.IPNetwork(cidr)
+            subnet, mask = ip.network.format(), ip.netmask.format()
+            next_hop = self._get_interface_next_hop_from_hosting_port(port)
+            conf_str = aci_asr1k_snippets.SET_TENANT_ROUTE_WITH_INTF % (
+                vrf_name, subnet, mask, out_itfc, next_hop)
+            self._edit_running_config(conf_str, 'SET_TENANT_ROUTE_WITH_INTF')
+
+    def _remove_tenant_net_route(self, ri, port):
+        cidr = port['subnets'][0]['cidr']
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py:137]: no need to check "if self._fullsync and (ri.router_id not in self._existing_cfg_dict['routes'])" similar to _add_tenant_net_route()?

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+                          'port': port}
+                raise cfg_exc.HAParamsMissingException(**params)
+
+    def _add_ha_hsrp(self, ri, port, is_external=False):
+        priority = None
+        if ri.router.get(ROUTER_ROLE_ATTR) in (ROUTER_ROLE_HA_REDUNDANCY,
+                                               ROUTER_ROLE_GLOBAL):
+            for router in ri.router[ha.DETAILS][ha.REDUNDANCY_ROUTERS]:
+                if ri.router['id'] == router['id']:
+                    priority = router[ha.PRIORITY]
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py:101]: break from the for loop after we got a match?

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py (hunk 1 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py]: Will fix.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py (hunk 2 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py]: Will fix.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py (hunk 3 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py:137]: For removals, this check isn't needed.

File: networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py (hunk 4 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py:101]: Good question -- this was leveraged from similar code in the stock ASR driver. I suspect that you're correct, but am going to leave this in for now... if a fix is needed, we'll create a separate patch that addresses any such issue across router type drivers.

**Comment by tbachman** [comment]
Pushed new version of patch, addressing reviewer comments.

**Pull request merged**